src/server/hotspot_partition_calculator.cpp

```diff
@@ -34,67 +35,84 @@ DSN_DEFINE_int64("pegasus.collector",
-            if (partition_data.total_qps - 1.00 > 0) {
-                data_samples.push_back(partition_data.total_qps);
-                table_qps_sum += partition_data.total_qps;
-                sample_count++;
+    dcheck_eq(_partitions_stat_histories.back().size(), _hot_points.size());
+    for (int data_type = 0; data_type <= 1; data_type++) {
+        // 0: READ_HOTSPOT_DATA; 1: WRITE_HOTSPOT_DATA
+        double table_qps_sum = 0, standard_deviation = 0, table_qps_avg = 0;
+        int sample_count = 0;
+        for (const auto &one_partition_stat_histories : _partitions_stat_histories) {
+            for (const auto &partition_stat : one_partition_stat_histories) {
+                if (partition_stat.total_qps[data_type] > 1.00) {
+                    table_qps_sum += partition_stat.total_qps[data_type];
+                    sample_count++;
+                }
             }
         }
-        temp_data.pop();
-    }
-    if (sample_count == 0) {
-        ddebug("_partition_stat_histories size == 0");
-        return;
-    }
-    table_qps_avg = table_qps_sum / sample_count;
-    for (const auto &data_sample : data_samples) {
-        standard_deviation += pow((data_sample - table_qps_avg), 2);
-    }
-    standard_deviation = sqrt(standard_deviation / sample_count);
-    const auto &anly_data = _partition_stat_histories.back();
-    for (int i = 0; i < _hot_points.size(); i++) {
-        double hot_point = (anly_data[i].total_qps - table_qps_avg) / standard_deviation;
-        // perf_counter->set can only be unsigned __int64
-        // use ceil to guarantee conversion results
-        hot_point = ceil(std::max(hot_point, double(0)));
-        _hot_points[i]->set(hot_point);
+
+        if (sample_count <= 1) {
+            ddebug("_partitions_stat_histories size <= 1");
+            return;
+        }
+        table_qps_avg = table_qps_sum / sample_count;
+        for (const auto &one_partition_stat_histories : _partitions_stat_histories) {
+            for (const auto &partition_stat : one_partition_stat_histories) {
+                if (partition_stat.total_qps[data_type] > 1.00) {
+                    standard_deviation +=
+                        pow((partition_stat.total_qps[data_type] - table_qps_avg), 2);
+                }
+            }
+        }
+        standard_deviation = sqrt(standard_deviation / (sample_count - 1));
+        const auto &anly_data = _partitions_stat_histories.back();
+        for (int i = 0; i < _hot_points.size(); i++) {
+            double hot_point = 0;
+            if (standard_deviation != 0) {
+                hot_point =
+                    (anly_data[i].total_qps[data_type] - table_qps_avg) / standard_deviation;
+            }
+            // perf_counter->set can only be unsigned uint64_t
+            // use ceil to guarantee conversion results
+            hot_point = ceil(std::max(hot_point, double(0)));
+            _hot_points[i][data_type]->get()->set(hot_point);
+        }
```

Comment:
I think this `data_analyse` can be totally refactored.

This function will firstly calculate qps statistics, then update hotpoint according to the `_partitions_stat_histories.back()`, you can seperate those two functions to make `data_analyse` clearer, such as:
```
void hotspot_partition_calculator::data_analyse()
{
     // your original dcheck_eq();
     for (int data_type = 0; data_type <= 1; data_type++) {
          double standard_deviation = 0, table_qps_avg = 0;
          if(get_hotpot_statistics(data_type, table_qps_avg, standard_deviation))
          {
              update_hot_point(_partitions_stat_histories.back(), data_type, table_qps_avg, standard_deviation);
          }
     }
     
}

bool hotspot_partition_calculator::get_hotpot_statistics(partition_qps_type data_type, double &qps_avg, 
double &standard_deviation)
{
     // calculate qps_avg and standard_deviation
     // if `sample_count <= 1`, there is no need to calculate statistics, return false
     // otherwise return true 
}

void hotspot_partition_calculator::update_hot_point(std::vector<hotspot_partition_stat> cur_stat, 
partition_qps_type data_type, double qps_avg, double standard_deviation)
{
     // update hot points
}
```
This is a simple refaction suggestion, you can also raise your better solution to make code clearer.